File: matlab/metasignal_mat/tutorials/03_statistical_inference.m
%% Tutorial 3 — Statistical Inference
% Bootstrap confidence intervals, permutation tests, and group-level
% summary using perform_ttest from the metasignal_mat helper functions.
%
% > Speed note: meta-d', meta-noise, and meta-uncertainty each require an
% > MLE optimisation (~1-3 s per subject).  This tutorial uses the four
% > fast non-MLE measures (d', AUC2, Gamma, Phi) so it runs in seconds.
% > The concepts transfer directly to any of the 20 measures.

%% Setup
clear; close all; clc

root_dir = fileparts(fileparts(mfilename('fullpath')));
addpath(genpath(fullfile(root_dir, 'helperFunctions')));

nRatings   = 4;
n_trials   = 200;
n_subjects = 20;
accuracy   = 0.78;
rng(0);

FAST_LABELS = {"d'", 'AUC2', 'Gamma', 'Phi'};

%% 1. Simulate a group of n_subjects participants
participants = cell(n_subjects, 1);
for i = 1:n_subjects
    participants{i} = sim_participant(i, accuracy, n_trials, nRatings);
end
fprintf('Simulated %d participants\n', n_subjects);

%% 2. Group-level summary
individual = NaN(n_subjects, 4);
for i = 1:n_subjects
    pt = participants{i};
    individual(i,:) = fast_measures(pt.stim, pt.resp, pt.conf, nRatings);
end

fprintf('\n%-8s  %8s  %8s\n', 'Measure', 'Mean', 'SEM');
fprintf('%s\n', repmat('-', 1, 28));
for k = 1:4
    col  = individual(:,k);
    mu   = nanmean(col);
    sem  = nanstd(col,0) / sqrt(sum(~isnan(col)));
    fprintf('%-8s  %8.3f  %8.3f\n', FAST_LABELS{k}, mu, sem);
end

%% 3. Bootstrap confidence interval
% Resample trials with replacement and recompute the measure.
% The percentile interval of the resampled distribution is the CI.

pt1 = participants{1};
auc2_fn = @(s,r,c) SDTtype2AUC(s, r, c, nRatings);

ci_auc2 = bootstrap_ci(pt1.stim, pt1.resp, pt1.conf, auc2_fn, 2000, 0.95, 1);
fprintf('\nAUC2 95%% CI (participant 1): [%.3f, %.3f]\n', ci_auc2(1), ci_auc2(2));

fn_list = {@(s,r,c) compute_SDT_resp(s,r), ...
           @(s,r,c) SDTtype2AUC(s,r,c,nRatings), ...
           @(s,r,c) SDTgamma(s,r,c,nRatings), ...
           @(s,r,c) SDTphi(s,r,c,nRatings)};

fprintf('\n%-8s  %-25s\n', 'Measure', '95% CI');
fprintf('%s\n', repmat('-', 1, 36));
for k = 1:4
    ci = bootstrap_ci(pt1.stim, pt1.resp, pt1.conf, fn_list{k}, 2000, 0.95, k);
    fprintf('%-8s  [%.3f, %.3f]\n', FAST_LABELS{k}, ci(1), ci(2));
end

%% 4. Permutation test — two-condition comparison
% Pool all trials, shuffle condition labels n_perm times to build a null
% distribution of the difference in a measure between two conditions.

% Condition A: metacognitive (conf tracks accuracy)
ptA = sim_participant(1, 0.80, n_trials, nRatings);

% Condition B: same accuracy, random (non-metacognitive) confidence
rng(200);
stim_b  = randi([0,1], n_trials, 1);
resp_b  = stim_b;
flip_b  = rand(n_trials,1) > 0.80;
resp_b(flip_b) = 1 - resp_b(flip_b);
conf_b  = randi([1, nRatings], n_trials, 1);

[p_perm, obs_diff] = permutation_test(ptA.stim, ptA.resp, ptA.conf, ...
                                       stim_b, resp_b, conf_b, ...
                                       auc2_fn, 5000, 42);

fprintf('\nAUC2 observed difference (A - B): %.3f\n', obs_diff);
fprintf('Two-sided permutation p-value:    %.4f\n', p_perm);

%% 5. One-sample t-test across participants using perform_ttest
% perform_ttest wraps MATLAB's ttest and also returns Cohen's d and 95% CI.

auc2_vals = individual(:, 2);   % AUC2 for all 20 participants
delta_vs_chance = auc2_vals - 0.5;

fprintf('\nAUC2 vs chance (0.5):\n');
[p_t, t_stat, df, Cohen_d, CI] = perform_ttest(delta_vs_chance, 'AUC2 vs 0.5', 1);  %#ok<ASGLU>

fprintf('\n  Tip: call perform_ttest for any of the 20 measures, e.g.:\n');
fprintf('  perform_ttest(metas(:,1) - 0, ''meta-d'''' > 0'', 1)\n');

% =========================================================================
% Local helper functions  (must appear at the end of a MATLAB script file)
% =========================================================================

function pt = sim_participant(seed, acc, n, nR)
    saved = rng(seed);
    stim    = randi([0,1], n, 1);
    resp    = stim;
    flip    = rand(n,1) > acc;
    resp(flip) = 1 - resp(flip);
    correct = (stim == resp);
    conf    = zeros(n,1);
    conf( correct) = randi([3,  nR], sum( correct), 1);
    conf(~correct) = randi([1,   2], sum(~correct), 1);
    pt = struct('stim',stim,'resp',resp,'conf',conf);
    rng(saved);
end

function vals = fast_measures(stim, resp, conf, nRatings)
    try
        dp = compute_SDT_resp(stim, resp);
    catch
        vals = NaN(1,4); return;
    end
    auc2 = SDTtype2AUC(stim, resp, conf, nRatings);
    gam  = SDTgamma(stim, resp, conf, nRatings);
    phi  = SDTphi(stim, resp, conf, nRatings);
    vals = [dp, auc2, gam, phi];
end

function ci = bootstrap_ci(stim, resp, conf, measure_fn, n_boot, ci_level, seed)
    saved = rng(seed);
    n    = length(stim);
    vals = NaN(n_boot,1);
    for b = 1:n_boot
        idx = randi(n, n, 1);
        v   = measure_fn(stim(idx), resp(idx), conf(idx));
        if ~isnan(v), vals(b) = v; end
    end
    rng(saved);
    alpha = 1 - ci_level;
    ci = prctile(vals(~isnan(vals)), [100*alpha/2, 100*(1-alpha/2)]);
end

function [p_val, obs_diff] = permutation_test(stim_a, resp_a, conf_a, ...
                                               stim_b, resp_b, conf_b, ...
                                               measure_fn, n_perm, seed)
    saved    = rng(seed);
    obs_a    = measure_fn(stim_a, resp_a, conf_a);
    obs_b    = measure_fn(stim_b, resp_b, conf_b);
    obs_diff = obs_a - obs_b;

    all_stim = [stim_a; stim_b];
    all_resp = [resp_a; resp_b];
    all_conf = [conf_a; conf_b];
    n_a = length(stim_a);
    n   = length(all_stim);

    null = NaN(n_perm, 1);
    for t = 1:n_perm
        perm = randperm(n);
        ia   = perm(1:n_a);
        ib   = perm(n_a+1:end);
        d = measure_fn(all_stim(ia), all_resp(ia), all_conf(ia)) - ...
            measure_fn(all_stim(ib), all_resp(ib), all_conf(ib));
        if ~isnan(d), null(t) = d; end
    end
    rng(saved);
    null  = null(~isnan(null));
    if isempty(null), p_val = NaN; return; end
    p_val = mean(abs(null) >= abs(obs_diff));
end

File: docs/tutorials/live_scripts/03_statistical_inference.py
# %% [markdown]
# # Tutorial 3 — Statistical Inference
#
# Bootstrap confidence intervals, permutation tests, and group-level summaries.
#
# > **Speed note** — `bootstrap_measure` and `permutation_test` in
# > `metasignal.analysis` call `compute_all_measures` internally, which includes
# > slow MLE fits (~3–5 s each) for meta-d', metaNoise, and metaUncertainty.
# > This notebook instead demonstrates the same concepts using the fast
# > non-MLE functions directly so it runs in seconds. In production scripts where
# > runtime is less constrained, use the convenience API shown at the end.

# %% [markdown]
# ## Setup

# %%
import numpy as np
from scipy import stats
from scipy.stats import pearsonr
from metasignal import stdpy

n_ratings = 4

def simulate_participant(seed, accuracy=0.78, n_trials=200):
    r = np.random.default_rng(seed)
    stim    = r.choice([0, 1], n_trials)
    resp    = np.where(r.random(n_trials) < accuracy, stim, 1 - stim)
    correct = stim == resp
    conf    = np.where(
        correct, r.integers(3, n_ratings + 1, n_trials),
        r.integers(1, 3, n_trials),
    )
    return stim, resp, conf

# Fast non-MLE measures: compute directly without compute_all_measures
def fast_measures(stim, resp, conf):
    """Return (dprime, AUC2, Gamma, Phi) — all O(n), no MLE."""
    dprime, c, _ = stdpy.compute_sdt_resp(stim, resp)
    nr_s1, nr_s2 = stdpy.trials_to_counts(stim, resp, conf, n_ratings)
    auc2  = stdpy.compute_type2_auc(nr_s1, nr_s2)
    gamma = stdpy.compute_gamma(nr_s1, nr_s2)
    phi   = stdpy.compute_phi(nr_s1, nr_s2)
    return np.array([float(dprime), auc2, gamma, phi])

FAST_LABELS = ["d'", "AUC2", "Gamma", "Phi"]

participants = [simulate_participant(i) for i in range(20)]
print(f"Simulated {len(participants)} participants")

# %% [markdown]
# ## 1. Group-level summary

# %%
# Compute fast measures for every participant
individual = np.array([fast_measures(*p) for p in participants])
print("individual shape:", individual.shape)   # (20, 4)

mean   = np.nanmean(individual, axis=0)
sem    = np.nanstd(individual, axis=0, ddof=1) / np.sqrt(individual.shape[0])

print(f"\n{'Measure':<8} {'Mean':>8} {'SEM':>8}")
print("-" * 28)
for name, m, s in zip(FAST_LABELS, mean, sem):
    print(f"{name:<8} {m:8.3f} {s:8.3f}")

# %%
# metasignal.analysis.group_summary wraps this for all 20 measures.
# Production usage (slower due to MLE fits):
#
#   from metasignal.analysis import group_summary
#   summary = group_summary(participants, n_ratings=n_ratings)
#   print(summary["mean"])    # shape (20,)
#   print(summary["labels"])  # 20 measure names
print("See docstring: metasignal.analysis.group_summary")

# %% [markdown]
# ## 2. Bootstrap confidence interval
#
# Resample trials with replacement and recompute the measure each time.
# The percentile interval of the resampled values is the CI.

# %%
def bootstrap_ci(stim, resp, conf, measure_fn, n_boot=2000, ci=0.95, seed=0):
    """Percentile-bootstrap CI for one scalar measure."""
    rng = np.random.default_rng(seed)
    n   = len(stim)
    vals = []
    for _ in range(n_boot):
        idx = rng.integers(0, n, size=n)
        v   = measure_fn(stim[idx], resp[idx], conf[idx])
        if not np.isnan(v):
            vals.append(v)
    alpha = 1 - ci
    return float(np.percentile(vals, 100 * alpha / 2)), float(np.percentile(vals, 100 * (1 - alpha / 2)))

def auc2_fn(stim, resp, conf):
    nr_s1, nr_s2 = stdpy.trials_to_counts(stim, resp, conf, n_ratings)
    return stdpy.compute_type2_auc(nr_s1, nr_s2)

stim, resp, conf = participants[0]
lo, hi = bootstrap_ci(stim, resp, conf, auc2_fn, n_boot=2000)
print(f"AUC2 95% CI: [{lo:.3f}, {hi:.3f}]")

# %%
# CIs for all four fast measures
def gamma_fn(s, r, c):
    nr_s1, nr_s2 = stdpy.trials_to_counts(s, r, c, n_ratings)
    return stdpy.compute_gamma(nr_s1, nr_s2)

def phi_fn(s, r, c):
    nr_s1, nr_s2 = stdpy.trials_to_counts(s, r, c, n_ratings)
    return stdpy.compute_phi(nr_s1, nr_s2)

def dprime_fn(s, r, c):
    d, _, _ = stdpy.compute_sdt_resp(s, r)
    return d

fns = [dprime_fn, auc2_fn, gamma_fn, phi_fn]
print(f"{'Measure':<8} {'95% CI':<25}")
print("-" * 34)
for name, fn in zip(FAST_LABELS, fns):
    lo, hi = bootstrap_ci(stim, resp, conf, fn, n_boot=2000, seed=hash(name) % 1000)
    print(f"{name:<8}  [{lo:.3f}, {hi:.3f}]")

# %% [markdown]
# ## 3. Permutation test — two-condition comparison
#
# Pool all trials, shuffle condition labels `n_perm` times, measure the
# difference under each shuffle to build a null distribution.

# %%
def permutation_test(stim_a, resp_a, conf_a, stim_b, resp_b, conf_b,
                     measure_fn, n_perm=5000, seed=0):
    """Two-sided permutation test for a difference in one measure."""
    rng    = np.random.default_rng(seed)
    obs_a  = measure_fn(stim_a, resp_a, conf_a)
    obs_b  = measure_fn(stim_b, resp_b, conf_b)
    obs_diff = obs_a - obs_b

    all_stim = np.concatenate([stim_a, stim_b])
    all_resp = np.concatenate([resp_a, resp_b])
    all_conf = np.concatenate([conf_a, conf_b])
    n_a = len(stim_a)
    n   = len(all_stim)

    null = []
    for _ in range(n_perm):
        perm = rng.permutation(n)
        d = measure_fn(all_stim[perm[:n_a]], all_resp[perm[:n_a]], all_conf[perm[:n_a]]) -             measure_fn(all_stim[perm[n_a:]], all_resp[perm[n_a:]], all_conf[perm[n_a:]])
        if not np.isnan(d):
            null.append(d)

    p = float(np.mean(np.abs(null) >= abs(obs_diff))) if null else np.nan
    return p, obs_diff

# Condition A: metacognitive (conf tracks accuracy)
stim_a, resp_a, conf_a = simulate_participant(0, accuracy=0.80)

# Condition B: same accuracy, random confidence (no metacognition)
rng_b = np.random.default_rng(200)
stim_b = rng_b.choice([0, 1], 200)
resp_b = np.where(rng_b.random(200) < 0.80, stim_b, 1 - stim_b)
conf_b = rng_b.integers(1, n_ratings + 1, 200)

p_val, obs_diff = permutation_test(
    stim_a, resp_a, conf_a,
    stim_b, resp_b, conf_b,
    auc2_fn, n_perm=5000, seed=42,
)
print(f"AUC2 observed difference (A − B): {obs_diff:.3f}")
print(f"Two-sided p-value:                {p_val:.4f}")

# %% [markdown]
# ## 4. One-sample t-test across participants

# %%
auc2_vals = individual[:, 1]   # AUC2 for all 20 participants

t, p   = stats.ttest_1samp(auc2_vals[~np.isnan(auc2_vals)], popmean=0.5)
n_ok   = np.sum(~np.isnan(auc2_vals))
d      = t / np.sqrt(n_ok)

print(f"AUC2 vs chance (0.5): t({n_ok-1}) = {t:.3f}, p = {p:.4f}, Cohen's d = {d:.3f}")

# %% [markdown]
# ## Production API
#
# For full 20-measure results (including meta-d', metaNoise, metaUncertainty),
# use the convenience wrappers in `metasignal.analysis`. They are ideal for
# offline scripts with parallelism but will be slow in a notebook:
#
# ```python
# from metasignal.analysis import bootstrap_measure, permutation_test, group_summary
#
# # ~seconds per call — best run with multiprocessing for many participants
# lo, hi = bootstrap_measure(stim, resp, conf, n_ratings=4,
#                            measure_index=0,   # meta-d'
#                            n_boot=2000)
#
# p, diff = permutation_test(stim_a, resp_a, conf_a,
#                            stim_b, resp_b, conf_b,
#                            n_ratings=4, measure_index=5,   # M-Ratio
#                            n_perm=5000)
#
# summary = group_summary(participants, n_ratings=4)
# ```
